### PR TITLE
TensorFlow example

### PR DESCRIPTION
**Pull request by juntai-zheng**
Added a sample script for TensorFlow use with MLflow for the blog post.

### CODE REVIEW

File: example/tutorial/tensorflow_example.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        predict_df = pyfunc.predict(df)
+        print(predict_df)
+    finally:
+        shutil.rmtree(temp)
+
+
+if __name__ == "__main__":
+    # The Estimator periodically generates "INFO" logs; make these logs visible.
+    tf.logging.set_verbosity(tf.logging.INFO)
+    tf.app.run(main=main)
```
Comment (smurching) [inline example/tutorial/tensorflow_example.py]: Nit: newline at EOF

File: example/tutorial/tensorflow_example.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+def main(argv):
+    # Builds, trains and evaluates a tf.estimator. Then, exports it for inference, logs the exported model 
+    # with MLflow, and loads the fitted model back as a PyFunc to make predictions.
+    (x_train, y_train), (x_test, y_test) = tf.keras.datasets.boston_housing.load_data()
+    feat_cols = [tf.feature_column.numeric_column(key="features", shape=(13,))]
+    feat_spec = {"features":tf.placeholder("float", name="features", shape=[102, 13])}
+    regressor = tf.estimator.DNNRegressor(
+    hidden_units=[50, 20],
+    feature_columns=feat_cols)
+    train_input_fn = tf.estimator.inputs.numpy_input_fn({"features": x_train}, y_train, batch_size=404, num_epochs=None, shuffle=True)
```
Comment (smurching) [inline example/tutorial/tensorflow_example.py]: Sorry, I randomly picked a batch size of 404 in the original script. Does it still give good results if we don't pass the batch size?

File: example/tutorial/tensorflow_example.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        tracking.start_run()
+        saved_estimator_path = regressor.export_savedmodel(temp, receiver_fn).decode("utf-8")
+        # Logging the saved model
+        tensorflow.log_saved_model(saved_model_dir=saved_estimator_path, signature_def_key="predict", artifact_path="model")
+        # Reloading the model
+        pyfunc = tensorflow.load_pyfunc(saved_estimator_path)
+        df = pd.DataFrame(data=x_test, columns=["features"] * 13)
+        # Predicting on the loaded Python Function
+        predict_df = pyfunc.predict(df)
+        print(predict_df)
```
Comment (smurching) [inline example/tutorial/tensorflow_example.py]: Would be nice to construct a new DataFrame containing two columns: the single output column in  pyfunc.predict() and the original labels, and print that here.

File: example/tutorial/tensorflow_example.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import tempfile
+import tensorflow as tf
+
+
+def main(argv):
+    # Builds, trains and evaluates a tf.estimator. Then, exports it for inference, logs the exported model 
+    # with MLflow, and loads the fitted model back as a PyFunc to make predictions.
+    (x_train, y_train), (x_test, y_test) = tf.keras.datasets.boston_housing.load_data()
+    feat_cols = [tf.feature_column.numeric_column(key="features", shape=(13,))]
+    feat_spec = {"features":tf.placeholder("float", name="features", shape=[102, 13])}
```
Comment (smurching) [inline example/tutorial/tensorflow_example.py]: Can we leave the leading dimension of the placeholder unspecified? i.e. `None`?

**Review by smurching — COMMENTED** [review]
Thanks @juntai-zheng, this generally looks good barring a few small comments

File: example/tutorial/tensorflow_example.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    train_input_fn = tf.estimator.inputs.numpy_input_fn({"features": x_train}, y_train, batch_size=404, num_epochs=None, shuffle=True)
+    regressor.train(train_input_fn, steps=1000)
+    test_input_fn = tf.estimator.inputs.numpy_input_fn({"features": x_test}, y_test, batch_size=404, num_epochs=None, shuffle=True)
+    # Compute mean squared error
+    regressor.evaluate(test_input_fn, steps=100)
+    # Building a receiver function for exporting
+    receiver_fn = tf.estimator.export.build_raw_serving_input_receiver_fn(feat_spec)
+    temp = tempfile.mkdtemp()
+    try:
+        tracking.start_run()
```
Comment (mparkhe) [inline example/tutorial/tensorflow_example.py]: Can we move this `tracking.start_run()` call further up to include training phase? Perhaps log some params and metrics?

File: example/tutorial/tensorflow_example.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import shutil
+import tempfile
+import tensorflow as tf
+
+
+def main(argv):
+    # Builds, trains and evaluates a tf.estimator. Then, exports it for inference, logs the exported model 
+    # with MLflow, and loads the fitted model back as a PyFunc to make predictions.
+    (x_train, y_train), (x_test, y_test) = tf.keras.datasets.boston_housing.load_data()
+    feat_cols = [tf.feature_column.numeric_column(key="features", shape=(13,))]
```
Comment (mparkhe) [inline example/tutorial/tensorflow_example.py]: Using these magic #s is scary. Can we figure this out from dataset? The concern is that if the shape of this data changes, this example will break.

File: example/tutorial/tensorflow_example.py (hunk 2 above)
Comment (juntai-zheng) [reply example/tutorial/tensorflow_example.py]: Avg. loss for batch size of 404 is 30, avg. loss for default batch size is 34. Do you want to remove the batch size in this case?

Comment (smurching) [reply example/tutorial/tensorflow_example.py]: Yeah I think that's a small enough difference that we can remove it, thanks :)

File: example/tutorial/tensorflow_example.py (hunk 6 above)
Comment (juntai-zheng) [reply example/tutorial/tensorflow_example.py]: As per @smurching's comment above, I was able to change the `102` to `None`. The 13 is based off of the number of feature columns used for training/testing/inference.

File: example/tutorial/tensorflow_example.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import tempfile
+import tensorflow as tf
+
+
+def main(argv):
+    # Builds, trains and evaluates a tf.estimator. Then, exports it for inference, logs the exported model 
+    # with MLflow, and loads the fitted model back as a PyFunc to make predictions.
+    (x_train, y_train), (x_test, y_test) = tf.keras.datasets.boston_housing.load_data()
+    # There are 13 features we are using for inference.
+    feat_cols = [tf.feature_column.numeric_column(key="features", shape=(13,))]
```
Comment (smurching) [inline example/tutorial/tensorflow_example.py]: I think you can replace 13 with `x_train.shape[1]` as per @mparkhe's suggestion

**Review by smurching — COMMENTED** [review]
One more comment, otherwise LGTM

File: example/tutorial/tensorflow_example.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    (x_train, y_train), (x_test, y_test) = tf.keras.datasets.boston_housing.load_data()
+    # There are 13 features we are using for inference.
+    feat_cols = [tf.feature_column.numeric_column(key="features", shape=(x_train.shape[1],))]
+    feat_spec = {"features":tf.placeholder("float", name="features", shape=[None, x_train.shape[1]])}
+    regressor = tf.estimator.DNNRegressor(
+    hidden_units=[50, 20],
+    feature_columns=feat_cols)
+    train_input_fn = tf.estimator.inputs.numpy_input_fn({"features": x_train}, y_train, num_epochs=None, shuffle=True)
+    tracking.start_run()
+    mlflow.log_param("Hidden Units", [50,20])
```
Comment (smurching) [inline example/tutorial/tensorflow_example.py]: Sorry one more comment - instead of hardcoding hidden units & steps could we store the values in variables & pass them to the `log_param` & `regressor.train` calls etc?

**Review by smurching — COMMENTED** [review]
Just one last comment, thanks for the work on this @juntai-zheng

**Review by smurching — APPROVED** [review]
LGTM thanks @juntai-zheng :)

File: example/tutorial/tensorflow_example.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    # Builds, trains and evaluates a tf.estimator. Then, exports it for inference, logs the exported model 
+    # with MLflow, and loads the fitted model back as a PyFunc to make predictions.
+    (x_train, y_train), (x_test, y_test) = tf.keras.datasets.boston_housing.load_data()
+    # There are 13 features we are using for inference.
+    feat_cols = [tf.feature_column.numeric_column(key="features", shape=(x_train.shape[1],))]
+    feat_spec = {"features":tf.placeholder("float", name="features", shape=[None, x_train.shape[1]])}
+    hidden_units = [50, 20]
+    steps = 1000
+    regressor = tf.estimator.DNNRegressor(
+    hidden_units=hidden_units,
```
Comment (mparkhe) [inline example/tutorial/tensorflow_example.py]: **super nit pick** : indentation

File: example/tutorial/tensorflow_example.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    feature_columns=feat_cols)
+    train_input_fn = tf.estimator.inputs.numpy_input_fn({"features": x_train}, y_train, num_epochs=None, shuffle=True)
+    tracking.start_run()
```
Comment (mparkhe) [inline example/tutorial/tensorflow_example.py]: recommend using

```
with tracking.start_run() as tracked_run:
  mlflow.log_param(...)
```

This will automatically call `end_run`

**Comment by mateiz** [comment]
Is this ready to merge? @juntai-zheng I think you can merge it directly since you got the LGTM from Sid.

**Comment by smurching** [comment]
Yep merging, thanks @juntai-zheng @mateiz @mparkhe !
